probe-rs/src/architecture/arm/component/mod.rs

```diff
@@ -90,126 +110,146 @@ fn find_component(
+                // Set EnFCont for continuous formatting even over SWO.
+                tpiu.set_formatter(0x102)?;
+            } else {
+                // Clear EnFCont to only pass through raw ITM/DWT data.
+                tpiu.set_formatter(0x100)?;
+            }
         }
-    } else {
-        log::warn!("SWO component not found - assuming TPIU-only configuration");
-    }
 
-    // Enable all ports of any trace funnels found.
-    for trace_funnel in components
-        .iter()
-        .filter_map(|comp| comp.find_component(PeripheralType::TraceFunnel))
-    {
-        let mut funnel = TraceFunnel::new(interface, trace_funnel);
-        funnel.unlock()?;
-        funnel.enable_port(0xFF)?;
-    }
+        TraceSink::Swo(config) => {
+            let mut swo = Swo::new(interface, find_component(components, PeripheralType::Swo)?);
+            swo.unlock()?;
 
-    // Configure ITM
-    let mut itm = Itm::new(interface, find_component(components, PeripheralType::Itm)?);
-    itm.unlock()?;
-    itm.tx_enable()?;
+            let prescaler = (config.tpiu_clk() / config.baud()) - 1;
+            swo.set_prescaler(prescaler)?;
 
-    // Configure DWT
-    let mut dwt = Dwt::new(interface, find_component(components, PeripheralType::Dwt)?);
-    dwt.enable()?;
-    dwt.enable_exception_trace()?;
+            match config.mode() {
+                SwoMode::Manchester => swo.set_pin_protocol(1)?,
+                SwoMode::Uart => swo.set_pin_protocol(2)?,
+            }
+        }
+
+        TraceSink::TraceMemory => {
+            let mut tmc = TraceMemoryController::new(
+                interface,
+                find_component(components, PeripheralType::Tmc)?,
+            );
+
+            // Clear out the TMC FIFO before initiating the capture.
+            tmc.disable_capture()?;
+            while !tmc.ready()? {}
+
+            // Configure the TMC for software-polled mode, as we will read out data using the debug
+            // interface.
+            tmc.set_mode(tmc::Mode::Software)?;
+
+            tmc.enable_capture()?;
+        }
+    }
 
-    // TODO: Replace flush
-    //interface.flush()
     Ok(())
 }
 
-/// Sets up all vendor specific bit of all the SWV components.
-///
-/// Expects to be given a list of all ROM table `components` as the second argument.
-fn setup_swv_vendor(
+pub(crate) fn read_trace_memory(
     interface: &mut Box<dyn ArmProbeInterface>,
     components: &[CoresightComponent],
-    config: &SwoConfig,
-) -> Result<(), Error> {
-    if components.is_empty() {
-        return Err(Error::architecture_specific(RomTableError::NoComponents));
+) -> Result<Vec<u8>, Error> {
+    let mut tmc =
+        TraceMemoryController::new(interface, find_component(components, PeripheralType::Tmc)?);
+
+    // TODO: In the future, it may be possible to dynamically read from trace memory
+    // without waiting for the FIFO to fill first.
+    let fifo_size = tmc.fifo_size()?;
+    while !tmc.full()? {
```

Comment:
Why wait until it's full, what if it never fills up completely or has a remainder in the end?

Reply:
In this configuration, there will never be remainder because of the flush mechanism that follows.

Right now, we're waiting for it to fill to guarantee that there's fully formatted frames available for de-multiplexing. If we didn't do this, we couldn't guarantee that the data stream wouldn't break between multiple calls to `read_trace_data()`.

Let me take a look at this again to see if it's reasonable to refactor it to support continuous operation.

Reply:
I've now refactored this to be more vesatile - you can now continuously call `read_trace_data()` with the sink set to TraceMemory, and you no longer require the FIFO to be full.

I've capped the amount of data that can be read in a single call to the size of the FIFO, since it's possible that trace data would be generated faster than it can be read out, which would result in infinite lockup.